septentrion/runner.py

```diff
@@ -14,124 +12,55 @@ class SQLRunnerException(Exception):
-        cursor.execute(statements)
-
-
-class MetaBlock(Block):
-    def __init__(self, command: str):
-        super(MetaBlock, self).__init__()
-        self.command = command
-        if command != "do-until-0":
-            raise SQLRunnerException("Unexpected command {}".format(command))
-
-    def run(self, cursor: Cursor) -> int:
-        total_rows = 0
-        # Simply call super().run in a loop...
-        delta = 0
-        batch_delta = -1
-        while batch_delta != 0:
-            batch_delta = 0
-            logger.debug("Running one block in a loop")
-            delta = super(MetaBlock, self).run(cursor)
-            if delta > 0:
-                total_rows += delta
-                batch_delta = delta
-            logger.debug("Batch delta done : %s", batch_delta)
-        return total_rows
-
-
-class Script(object):
+class Script:
     def __init__(
         self,
         settings: configuration.Settings,
         file_handler: Iterable[str],
         path: pathlib.Path,
     ):
-        file_lines = list(file_handler)
-        is_manual = files.is_manual_migration(
-            migration_path=path, migration_contents=file_lines
-        )
         self.settings = settings
-        if is_manual:
-            self.block_list = [Block()]
-        elif self.contains_non_transactional_keyword(file_lines):
-            self.block_list = [Block()]
-        else:
-            self.block_list = [SimpleBlock()]
-        for line in file_lines:
-            if line.startswith("--meta-psql:") and is_manual:
-                self.block_list[-1].close()
-                command = line.split(":")[1].strip()
-                if command == "done":
-                    # create a new basic block
-                    self.block_list.append(Block())
-                else:
-                    # create a new meta block
-                    self.block_list.append(MetaBlock(command))
-            else:
-                self.block_list[-1].append_line(line)
-        self.block_list[-1].close()
+        self.file_lines = list(file_handler)
+        self.path = path
 
     def run(self, connection):
-        with connection.cursor() as cursor:
-            for block in self.block_list:
-                block.run(cursor)
-
-    def contains_non_transactional_keyword(self, file_lines: Iterable[str]) -> bool:
-        keywords = self.settings.NON_TRANSACTIONAL_KEYWORD
-        for line in file_lines:
-            for kw in keywords:
-                if kw.lower() in line.lower():
-                    return True
-
-        return False
+        if any('--meta-psql:' in line for line in self.file_lines):
+            self._run_with_meta_loop()
+        else:
+            self._run_simple()
+
+    def _run_simple(self):
+        creds = []
+        for name in ['HOST', 'PORT', 'DBNAME', 'USERNAME', 'PASSWORD']:
+            value = getattr(self.settings, name)
+            if value:
+                creds += ['--' + name.lower(), value]
```

Comment:
wouldn't it be much clear to pass all those via environment ?

Reply:
If I define env vars, I must also pass mandatory env vars like PATH, otherwise psql won't be found. I guess I could do something like `env={'PGHOST': self.settings.HOST [...], **os.environ}`, but I don't know if it's a good practice.

Reply:
Yes, I was thinking of doing that. Is there any reason this wouldn't be a good practice ?

(I really wish subprocess had an "additional_env" or "env_override", but eh).

Reply:
psql will prioritize the cmd arguments over the env vars. I'm not sure of the priority in Septentrion, but passing explicit values to psql will ensure that we don't end up in a situation were Septentrion connects to a DB, but psql launches the migrations in another one.

Reply:
not convinced but this can be dealt with later.